tchannel/zipkin/tracers.py

```diff
@@ -135,23 +139,23 @@ def __init__(self, tchannel):
             A tchannel instance to send the trace info to zipkin server
         """
 
-        from ..thrift import client_for
-
-        # TODO add different thrift service name
-        TCollectorClient = client_for('tcollector', TCollector)
-
         self._tchannel = tchannel
-        self.client = TCollectorClient(self._tchannel)
 
-    def record(self, traces):
-        def submit_callback(f):
-            if f.exception():
-                log.error('Fail to submit zipkin trace',
-                          exc_info=f.exc_info())
+    def submit_callback(self, f):
+        if f.exception():
+            log.error(
+                'Fail to submit zipkin trace',
+                exc_info=f.exc_info()
+            )
 
+    def record(self, traces):
         for (trace, annotations) in traces:
-            f = self.client.submit(thrift_formatter(trace, annotations))
-            f.add_done_callback(submit_callback)
+            client = TCollectorClient(
+                self._tchannel,
+                protocol_headers={'shardKey': i64_to_base64(trace.trace_id)}
+            )
+            f = client.submit(thrift_formatter(trace, annotations))
+            f.add_done_callback(self.submit_callback)
```

Comment:
So, if I understand correctly, the sole reason submit_callback is a public method is so that you can override it and get the real result for a test. That's not a very good approach to solve this problem.

What if you returned those futures as a list (in addition to adding the `submit_callback` for logging -- which doesn't have too be a public method) from the `record` call? The caller of record can choose to yield those futures and get the results if necessary. (You can yield a list of futures to get a list of the results.)

Generally, if you're making an asynchronous call, it is beneficial if your caller can wait for/get the result.